registry/spec.md

```diff
@@ -1483,165 +1624,1718 @@ An attempt to delete all versions MUST generate an error.
+  and documents all versions of all Protobuf schemas that are used in the
+  directory.
+- The [Message Definitions Registry](#message-definitions-registry) section
+  describes the metadata description of events and messages. The payload schemas
+  for events and messages can be embedded in the definition, reference an
+  external schema document, or can be referenced into the schema registry. The
+  message definitions registry is universally applicable to any asynchronous
+  messaging and eventing scenario. You might define a group of definitions that
+  describe precisely which messageds with which metadata are permitted to flow
+  into a channel and can thus be expected by consumers of that channel and then
+  associate that definition group with a topic or queue in your eventing or
+  messaging infrastructure. That association might be a metadata attribute on
+  the topic or queue in the messaging infrastructure that embeds the metadata or
+  points to it.
+- The [Endpoint Registry](#endpoint-registry) section defines the metadata
+  description of network endpoints that accept or emit events and messages. The
+  endpoint registry is a formal description of associations of message
+  definitions and network endpoints, which can be used to discover endpoints
+  that consume or emit particular messages or events via a central registry. The
+  message definitions can be embedded into the endpoint metadata or as
+  a reference into the message definitions registry.
+
+The metadata model is structured such that network endpoint information and
+message metadata and payload schemas can be described compactly in a single
+object in the simplest case or can be spread out and managed across separate
+registry products in a sophisticated large-enterprise scenario.
+
+The following is an exemplary, compact definition of an MQTT 5.0 consumer
+endpoint with a single, embedded message definition using an embedded Protobuf 3
+schema for its payload.
+
+``` JSON
+{
+    "$schema": "https://cloudevents.io/schemas/registry",
+    "specversion": "0.3-wip",
+    "id": "urn:uuid:3978344f-8596-4c3a-a978-8fc9a6a469f7",
+    "endpoints" : 
+    {
+       "com.example.telemetry" : {
+           "type": "endpoint",
+           "id" : "com.example.telemetry",
+            "usage": "consumer",
+            "config": {
+                "protocol": "MQTT/5.0",
+                "strict": false,
+                "endpoints": [
+                    "mqtt://mqtt.example.com:1883",
+                ],
+                "options" : {
+                    "topic": "{deviceid}/telemetry"
+                }
+            },
+            "format" : "CloudEvents/1.0"
+            "definitions": {
+                  "com.example.telemetry" : {
+                    "type": "cloudevent",
+                    "id": "com.example.telemetry",
+                    "description": "device telemetry event",
+                    "format": "CloudEvents/1.0",
```

Comment:
What is "type"? I don't see that defined, but maybe I'm missing it. If it remains, it might be hard for people to easily understand why we need it and "format" when they're both variants of "cloudevent"

Reply:
OpenAPI has introduced its own sub-flavor of JSON schema for schemas that use inheritance as we do and that requires a "discriminator" 

https://spec.openapis.org/oas/latest.html#discriminator-object

The unfortunate effect of "discriminator" is that every instance of every derived type needs to labeled in some way that the type can be inferred. 

Since we need to use OpenAPI for the REST API, I cannot avoid buying into that scheme.

In our schema doc (schemas/ce_registry_doc.json in my PR), I have started to work around the "type" field and use "format" as the discriminator. That doesn't make me happy, but is an effect orf OpenAPI imposing that requirement and us needing OpenAPI:

```json
 "discriminator": {
          "propertyName": "format",
          "mapping": {
            "AMQP/1.0": "#/components/schemas/amqpDefinition10",
            "AMQP": "#/components/schemas/amqpDefinition",
            "MQTT/3.1.1": "#/components/schemas/mqttDefinition311",
            "MQTT/5.0": "#/components/schemas/mqttDefinition50",
            "MQTT": "#/components/schemas/mqttDefinition",
            "HTTP/1.1": "#/components/schemas/httpDefinition11",
            "HTTP": "#/components/schemas/httpDefinition",
            "HTTP/2": "#/components/schemas/httpDefinition2",
            "HTTP/3": "#/components/schemas/httpDefinition3",
            "KAFKA/0.11": "#/components/schemas/kafkaDefinition011",
            "KAFKA": "#/components/schemas/kafkaDefinition",
            "CloudEvents/1.0": "#/components/schemas/cloudEventDefinition10",
            "CloudEvents": "#/components/schemas/cloudEventDefinition",
            "Generic": "#/components/schemas/genericDefinition"
          }

```

`type` will have to remain for all derivatives of `resource` (including `definition`), but with leaning on `format` I will be able to avoid further subdivision of `definition`

``` json
 "resource": {
        "type": "object",
        "discriminator": {
          "propertyName": "type",
          "mapping": {
            "definitiongroup": "#/components/schemas/definitiongroup",
            "definition": "#/components/schemas/definition",
            "schemagroup": "#/components/schemas/schemaGroup",
            "endpoint": "#/components/schemas/endpoint",
            "schema": "#/components/schemas/schema",
            "schemaversion": "#/components/schemas/schemaVersion"
          }
        },
```